### PR TITLE
Profile Picture added on API and Backend

### PR DESCRIPTION
**Pull request by aksinha-nerdapplabs**
PR for [Issue# 5](https://github.com/nerdapplabs/authOauth/issues/5)

### CODE REVIEW

File: src/ApiBundle/Controller/Admin/UserController.php [hunk 1]
```diff
@@ -83,6 +102,10 @@ public function newAction(Request $request)
               $flashMsg = $this->get('translator')->trans('flash.user_created_successfully');
               $this->addFlash('success', $flashMsg);
 
+            } catch(UploadException $e) {
+              $this->logMessage(400, 'danger', $e->getMessage());
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/Admin/UserController.php]: should not we use logMessageAndFlash?

File: src/ApiBundle/Controller/Admin/UserController.php [hunk 2]
```diff
@@ -162,6 +212,10 @@ public function editAction(User $user, Request $request)
 
             // Always catch exact exception for which flash message or logger is needed,
             // otherwise catch block will not get executed on higher or lower ranked exceptions.
+            } catch(UploadException $e) {
+              $this->logMessage(400, 'danger', $e->getMessage());
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/Admin/UserController.php]: logMessageAndFlash?

File: src/ApiBundle/Controller/AuthController.php [hunk 3]
```diff
@@ -223,11 +230,12 @@ private function validateAdminUser(Request $request) {
       *      {"name"="dob", "dataType"="datetime", "required"=true, "description"="date of birth mm/dd/yyyy"},
       *      {"name"="email", "dataType"="email", "required"=true, "description"="Email"},
       *      {"name"="email_confirmation", "dataType"="integer", "required"=true, "description"="0-email confirmation not required, 1-required"},
+      *      {"name"="image", "dataType"="image/jpeg, image/jpg, image/gif, image/png", "required"=false, "description"="Profile Picture within 1024k size"},
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: where did we limit 1024K

File: src/ApiBundle/Entity/User.php [hunk 4]
```diff
@@ -48,6 +49,17 @@ class User extends BaseUser
     protected $dob;
 
     /**
+    * @var string $image
+    * @Assert\File(
+    *     maxSize = "1024k",
```
Comment (praveen-garg) [inline src/ApiBundle/Entity/User.php]: Okay, here it is! cool!

**Review by praveen-garg — CHANGES_REQUESTED** [review]
minor comments;

File: src/ApiBundle/Entity/User.php (hunk 4 above)
Comment (praveen-garg) [reply src/ApiBundle/Entity/User.php]: As discussed, User (who is signing up) should be able add/ remove or update user profile image; not the admin only.

File: src/ApiBundle/Form/UserType.php [hunk 5]
```diff
@@ -32,7 +32,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('email', EmailType::class)
             ->add('dob', DateType::class, array('widget' => 'single_text', 'format' => 'M/d/y'))
             ->add('username', TextType::class)
-            ->add('password', PasswordType::class, array('data' => ''))
+            ->add('password', TextType::class)
```
Comment (praveen-garg) [inline src/ApiBundle/Form/UserType.php]: changed it back to TextType? 
it was chnaged to PasswordType in last PR
https://github.com/nerdapplabs/authOauth/pull/9/files#diff-3683af68b1a14d46bf30b457a8ec4ddfR31

**Comment by aksinha-nerdapplabs** [comment]
@praveen-garg have a look. Changes are final till now.

File: src/ApiBundle/Controller/AuthController.php [hunk 6]
```diff
@@ -252,22 +260,23 @@ public function postRegisterAction(Request $request)
...
+            $user->setImage($newFile);
         }
 
         $user->setUsername($request->request->get('username'));
         $user->setPlainPassword($request->request->get('password'));
+        $user->setPassword($request->request->get('password'));
         $user->setEmail($request->request->get('email'));
         $user->setFirstname($request->request->get('firstname'));
-        $user->setLastname($request->request->get('lastname'));
+        $user->setLastname($request->request->get('lastname') ? $request->request->get('lastname') : null);
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: should not we make it EMPTY string ? ""

` $user->setLastname($request->request->get('lastname') ? $request->request->get('lastname') : "");`

File: src/ApiBundle/Controller/AuthController.php [hunk 7]
```diff
@@ -282,8 +291,9 @@ public function postRegisterAction(Request $request)
         $oAuthRtn = 'Pending';
         $msg = 'N.A.';
         $grantType = 'password';
+        $loginConfirmation = $this->container->getParameter('login_confirmation');
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: what is the use of this var?

File: src/ApiBundle/Controller/AuthController.php [hunk 8]
```diff
@@ -282,8 +291,9 @@ public function postRegisterAction(Request $request)
...
 
-        if ('1' == $request->request->get('email_confirmation')) {
+        if (true == $this->container->getParameter('email_confirmation') ) {
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: email_confirmation > registration_requires_email_confirmation

File: src/ApiBundle/Controller/AuthController.php [hunk 9]
```diff
@@ -446,6 +467,11 @@ public function getProfilePicAction()
             $this->logAndThrowError(400, 'Invalid User', $this->get('translator')->trans('api.show_error_perm_show', array(), 'messages', $request->getLocale()), $request->getLocale());
         }
 
+        // If no image file, then return empty response
+        if (!$user->getImage()) {
+            return new Response();
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: can you please explain more? 

If no image file, then return empty response ?

File: src/ApiBundle/Controller/AuthController.php [hunk 10]
```diff
@@ -463,7 +489,7 @@ public function getProfilePicAction()
       *  resource=true,
       *  description="Fetch User profile detail. Access token to be provided in header (Authorization = Bearer <access token>)",
       *  parameters={
-      *      {"name"="image", "dataType"="image/jpeg, image/jpg, image/gif, image/png", "required"=false, "description"="Profile Picture within 1024k size"},
+      *      {"name"="image", "dataType"="image/jpeg, image/jpg, image/gif, image/png", "required"=false, "description"="jpeg/jpg/png/gif Pic Size: 1024k, Width: 50-300px, Height: 50-300px"},
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: older looks better!

File: src/ApiBundle/Resources/config/config.yml [hunk 11]
```diff
@@ -2,8 +2,9 @@ parameters:
     # This parameter defines the codes of the locales (languages) enabled in the application
     # app_locales: en|fr|de|es|cs|nl|ru|uk|ro|pt_BR|pl|it|ja|id|ca|sl
     app_locales: en|fr|hi
-    images_profile_suffix: '/images/profile/'
-    images_profile_directory: '%kernel.root_dir%/../web%images_profile_suffix%'
+    images_profile_directory: '%kernel.root_dir%/../web/images/profile'
+    email_confirmation: false
+    login_confirmation: false
```
Comment (praveen-garg) [inline src/ApiBundle/Resources/config/config.yml]: see previous comments

**Comment by praveen-garg** [comment]
@aksinha-nerdapplabs  we must keep PR specific to the issue; not sure why this PR has  registration_requires_email_confirmation related change. a new PR (with minimum & to the point chnages) could have been easy to review/ test and merge

File: src/ApiBundle/Controller/AuthController.php (hunk 9 above)
Comment (aksinha-nerdapplabs) [reply src/ApiBundle/Controller/AuthController.php]: yeah, Empty with 200 ok.

File: src/ApiBundle/Controller/AuthController.php [hunk 12]
```diff
@@ -258,7 +292,11 @@ public function postRegisterAction()
         $msg = 'N.A.';
         $grantType = 'password';
 
-        if ('1' == $request->request->get('email_confirmation')) {
+        // If Login confirmation is set, then upon successful registration, user will be logged in automatically,
+        // else user will be redirected to a login page to log in with the new credentials.
+        $loginConfirmation = $this->container->getParameter('login_confirmation');
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/AuthController.php]: login_confirmation > login_required_after_signup

File: src/ApiBundle/Controller/UserController.php [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+     */
+    protected function authenticateUser(UserInterface $user, Response $response)
+    {
+        try {
+            $this->container->get('fos_user.security.login_manager')->loginUser(
+                $this->container->getParameter('fos_user.firewall_name'),
+                $user,
+                $response);
+        } catch (AccountStatusException $ex) {
+            // We simply do not authenticate users which do not pass the user
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/UserController.php:140]: but we should log?

File: src/ApiBundle/Controller/Admin/UserController.php [hunk 14]
```diff
@@ -53,10 +56,22 @@ public function newAction(Request $request)
...
 
         $form = $this->createForm(UserType::class, $user);
 
+        // Role added in admin area
+        $form->add('roles', CollectionType::class, array(
+                          'entry_type'   => ChoiceType::class,
+                          'entry_options'  => array(
+                              'label' => false,
+                              'choices'  => array(
+                                  'ROLE_ADMIN' => 'ROLE_ADMIN',
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/Admin/UserController.php:69]: ? yes?

File: src/ApiBundle/Controller/Admin/UserController.php [hunk 15]
```diff
@@ -127,10 +142,23 @@ public function editAction(User $user, Request $request)
...
         $deleteForm = $this->createDeleteForm($user);
         $locale = $request->getLocale();
 
+        // Role added in admin area
+        $editForm->add('roles', CollectionType::class, array(
+                          'entry_type'   => ChoiceType::class,
+                          'entry_options'  => array(
+                              'label' => false,
+                              'choices'  => array(
+                                  'ROLE_ADMIN' => 'ROLE_ADMIN',
```
Comment (praveen-garg) [inline src/ApiBundle/Controller/Admin/UserController.php:164]: are your sure?

**Review by praveen-garg — APPROVED** [review]
DRY review; need more testing..

**Pull request merged**